### PR TITLE
[#784] AMQP Adapter: Implement Command & Control

### PR DESCRIPTION
**Pull request by Alfusainey**
@sophokles73 This patch address #784, with two separate commits:
* 6b67f7b: allow clients to subscribe to commands + unit tests
* 86bb838: allow clients to publish command responses + unit tests

### CODE REVIEW

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java [hunk 1]
```diff
@@ -235,13 +248,7 @@ private void connectionRequestHandler(final ProtonConnection connRequest) {
...
         });
         connRequest.senderOpenHandler(sender -> {
-            HonoProtonHelper.setDefaultCloseHandler(sender);
-            // this should not happen -> no request-response for device clients.
-            LOG.debug("client [container: {}] wants to open a link [address: {}] for receiving messages",
-                    connRequest.getRemoteContainer(), sender.getRemoteSource());
-            sender.setCondition(ProtonHelper.condition(AmqpError.NOT_ALLOWED,
-                    "this adapter only forwards message to downstream applications"));
-            sender.close();
+            handleRemoteSenderOpen(sender, connRequest);
```
Comment (sysexcontrol) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: As done above already:
`
// when an Attach frame with "role = receiver" is received

// when an Attach frame with "role = sender" is received

Would improve the (already good) documentation of the AMQP handling here IMHO.
`

**Comment by Alfusainey** [comment]
> can we agree on first having a PR with the concept how we want to provide Command and Control for the AMQP Adapter?
This would clarify a lot of things, and we can come back to the implementation later...

@sysexcontrol Thank you! Yes, i think this makes a lot of sense. i will create a PR with the concepts next

> It might even make sense FMPOV to only have control in role receiver (for receiving commands) and control in role sender (for sending a response back), but this would need discussion on the concept level first.

Yes, i agree! I will include this in the concept description

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java [hunk 2]
```diff
@@ -352,6 +357,133 @@ protected void handleRemoteReceiverOpen(final ProtonReceiver receiver, final Pro
...
     }
 
+    /**
+     * This method is invoked when this server receives an AMQP Attach frame (from a device) to establish a link
+     * for receiving commands from downstream applications.
+     * 
+     * @param sender The link to use for sending commands to devices.
+     * @param connection The connection instance with the device.
+     */
+    protected void handleRemoteSenderOpen(final ProtonSender sender, final ProtonConnection connection) {
```
[the remark this answers is not in the document]
Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: ok. i will update this later once the concept doc is out of the way. thanks!

**Comment by sophokles73** [comment]
@sysexcontrol Are you going to continue reviewing this PR or shall I take over?

**Comment by sysexcontrol** [comment]
@sophokles73 : looking at my current to do list : please take over, thx!

**Comment by Alfusainey** [comment]
you might want to delay this review a bit since the C&C concepts changed a bit. I will update it as soon as I find free time

**Comment by Alfusainey** [comment]
@sophokles73 patch updated to reflect the new `C&C concepts`

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java [hunk 3]
```diff
@@ -446,14 +586,66 @@ protected void uploadMessage(final AmqpContext context) {
...
+            if (commandResponse == null) {
+                return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                        "malformed command response address"));
+            } else {
+                return sendCommandResponse(context.getTenantId(), commandResponse, null)
+                        .map(delivery -> {
+                            LOG.trace(
+                                    "successfully forwarded command response from device [tenant-id: {}, device-id: {}]",
+                                    context.getTenantId(), context.getDeviceId());
+                            if (context.isRemotelySettled()) {
```
[the remark this answers is not in the document]
Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: ok, I see!

what happens if the adapter is disabled for a tenant? should the adapter still receive commands from applications or allow devices send command responses?  in the case of publishing `events/telemetry`, the adapter responds with a `403` status code

[the remark this answers is not in the document]
Comment (sophokles73) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: If the adapter is disabled for a device then the device should not be able to connect in the first place, should it?

[the remark this answers is not in the document]
Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: > If the adapter is disabled for a device

I think you wanted to say for a `tenant`? If yes, then I will add this in the next commit (together with the other modifications)

[the remark this answers is not in the document]
Comment (sophokles73) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: sure

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java [hunk 4]
```diff
@@ -363,55 +364,84 @@ protected void handleRemoteReceiverOpen(final ProtonReceiver receiver, final Pro
...
-                        sender.setQoS(sender.getRemoteQoS());
-                        HonoProtonHelper.setCloseHandler(sender, remoteDetach -> {
-                            sendDisconnectedTtdEvent(tenantId, deviceId, authenticatedDevice);
-                            closeCommandConsumer(tenantId, deviceId);
-                            onLinkDetach(sender);
+                    if (!CommandConstants.isCommandEndpoint(validAddress.getEndpoint())) {
+                        closeLinkWithError(sender, new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                                "Only the command endpoint is supported"));
+                    } else {
+                        final Future<TenantObject> tenantTracker = getTenantConfiguration(validAddress.getTenantId(), null);
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: I do not think that this is the right place to do this. Why would we allow a device to connect to the AMQP adapter in the first place, if the AMQP adapter has been disabled for the device's tenant?

Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java]: yes that make sense. will update the patch

**Comment by Alfusainey** [comment]
@sophokles73: patch updated with 6761f43. now the test for disabled tenant happens during connection establishment

looks like the Java11 build fails due to a timeout. its not related to my changes but can take a look later..

File: adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorFactory.java [hunk 5]
```diff
@@ -200,7 +201,16 @@ private void verifyPlain(final byte[] saslResponse, final Handler<AsyncResult<De
                     completer.handle(Future.failedFuture(new CredentialException(
                             "username does not comply with expected pattern [<authId>@<tenantId>]")));
                 } else {
-                    getUsernamePasswordAuthProvider().authenticate(credentials, completer);
+                    getTenantObject(credentials.getTenantId()).map(tenant -> {
+                        if (tenant.isAdapterEnabled(Constants.PROTOCOL_ADAPTER_TYPE_AMQP)) {
+                            getUsernamePasswordAuthProvider().authenticate(credentials, completer);
```
Comment (sophokles73) [inline adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorFactory.java]: we should be able to do both in parallel:
* getting the Tenant
* getting the Credentials

by using a CompositeFuture. However, we can postpone that optimization til later ...

Comment (Alfusainey) [reply adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpAdapterSaslAuthenticatorFactory.java]: ok, I will take a look

**Comment by Alfusainey** [comment]
@sophokles73: done!

**Comment by sophokles73** [comment]
@Alfusainey what about adding an integration test for C&C?

**Comment by Alfusainey** [comment]
> @Alfusainey what about adding an integration test for C&C?

yes sure.. i was waiting for this to be merged into master..

**Pull request merged**